Autostart

berkeyvx opened the pull request:
Fixes #86

I did a little different than your idea. Glib functions require GFile, too much overhead for such simple implementation. So i decided to go with filesystem. 

```cpp
auto const CURRENT_DIR = std::string{g_get_current_dir()};
```
I use CURRENT_DIR to find desktop file. I decided to use this because snap installation and local installation directories are different. `g_get_current_dir` give us following directories.
* /snap/whatsapp-for-linux/.../bin `(this needs to be tested)`
* /usr/local/bin

I manipulate this locations here: 
```cpp
bool Settings::findDesktopFile(std::string& outPath)
{
    outPath = CURRENT_DIR;
    outPath.erase(CURRENT_DIR.length() - 3);
    outPath += "share/applications/whatsapp-for-linux.desktop";

    return fs::exists(outPath);
}
```
Function take empty string, using CURRENT_DIR it constructs `.desktop` file path, then checks if file is exists. It is a utility function for `setAutoStart()`.


If findDesktopFile fails to find `.desktop` file, we do nothing. We can change this part after snap installation test. Remain part is the same as i wrote before
```cpp
void Settings::setAutoStart()
{
    std::string path{};
    if(!Settings::instance().findDesktopFile(path))
        return;

    if(Settings::instance().getStartInTray())
        std::system(("cp " + path + " ~/.config/autostart/").c_str());
    else
        std::system("rm ~/.config/autostart/whatsapp-for-linux.desktop");
}
```

Tbh i don't like the `bool Settings::findDesktopFile(std::string& outPath)` function. I am open to advices.

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +159,24 @@ bool Settings::startInTray() const
+bool Settings::findDesktopFile(std::string& outPath)
+{
+    outPath = CURRENT_DIR;
+    outPath.erase(CURRENT_DIR.length() - 3);
```
Please use `std::string::size` since it's more convenient.

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +159,24 @@ bool Settings::startInTray() const
+
+bool Settings::findDesktopFile(std::string& outPath)
+{
+    outPath = CURRENT_DIR;
```
See my comment above.

eneshecan commented:
Thanks! Looks good except finding the desktop file. What happens when current dir is somehow not system folder such as `HOME/Projects/whatsapp-for-linux/build` which is possible in development environment. That's why I suggested to check folders `/usr/share/applications` and `usr/local/share/applications`. Snap also stores those files in app specific folders which I cannot recall. Just ignore snap for now.

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +159,24 @@ bool Settings::startInTray() const
+        std::system("rm ~/.config/autostart/whatsapp-for-linux.desktop");
+}
+
+bool Settings::findDesktopFile(std::string& outPath)
```
As outPath is an out variable you should also return the string. For that, you can use a `std::pair<bool, std::string>`

eneshecan commented:
Also, please resolve the conflicts before pushing another commit.

berkeyvx commented:
> Thanks! Looks good except finding the desktop file. What happens when current dir is somehow not system folder such as `HOME/Projects/whatsapp-for-linux/build` which is possible in development environment. That's why I suggested to check folders `/usr/share/applications` and `usr/local/share/applications`. Snap also stores those files in app specific folders which I cannot recall. Just ignore snap for now.

CURRENT_DIR returns path where exec file located. 
So in development environment it will return something like this
* HOME/Projects/whatsapp-for-linux/build/debug/ 

It will fail if you try it. But it will work for installation

That is why i think this approach is work with snap too. My snap exec is inside `/snap/whatsapp-for-linux/23/bin/`. I have'nt tested yet but i think CURRENT_DIR will point to that directory, from there i construct `/snap/whatsapp-for-linux/23/share/applications/whatsapp-for-linux.desktop`. with that function.

berkeyvx commented:
I'll make fixes on others as soon as possible. Thanks for feedback

eneshecan commented:
> > Thanks! Looks good except finding the desktop file. What happens when current dir is somehow not system folder such as `HOME/Projects/whatsapp-for-linux/build` which is possible in development environment. That's why I suggested to check folders `/usr/share/applications` and `usr/local/share/applications`. Snap also stores those files in app specific folders which I cannot recall. Just ignore snap for now.
> 
> CURRENT_DIR returns path where exec file located. 
> So in development environment it will return something like this
> * HOME/Projects/whatsapp-for-linux/build/debug/ 
> 
> It will fail if you try it. But it will work for installation
> 
> That is why i think this approach is work with snap too. My snap exec is inside `/snap/whatsapp-for-linux/23/bin/`. I have'nt tested yet but i think CURRENT_DIR will point to that directory, from there i construct `/snap/whatsapp-for-linux/23/share/applications/whatsapp-for-linux.desktop`. with that function. 

Hmmm, fair enough. We can keep this approach. Thanks a lot!

berkeyvx commented:
There was a problem with the last approach. This pr only contains solution for cmake installation. We still need to find a solution for snap. I am quite busy for next weeks. I will search a solution for snap when i am available

berkeyvx commented:
`/snap/<snapname>/current    - Symlink to current revision, if enabled` -> https://snapcraft.io/docs/system-snap-directory
`/snap/whatsapp-for-linux/current/share/applications/whatsapp-for-linux.desktop`

I think i can use this path. But i need to do tests. Is there a way to make changes then install using snap from development environment? @eneshecan

berkeyvx commented:
> `/snap/<snapname>/current - Symlink to current revision, if enabled` -> https://snapcraft.io/docs/system-snap-directory
> `/snap/whatsapp-for-linux/current/share/applications/whatsapp-for-linux.desktop`
> 
> I think i can use this path. But i need to do tests. Is there a way to make changes then install using snap from development environment? @eneshecan

I tested manually, it works

eneshecan commented on src/Settings.hpp:
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <glibmm/keyfile.h>
+#include <giomm/file.h>
```
This is redundant.

eneshecan commented on src/Settings.cpp:
```diff
@@ -2,6 +2,7 @@
 #include <sys/stat.h>
 #include <fstream>
 #include <iostream>
+#include <giomm/file.h>
```
You include the cpp api but use the c functions below. Please use the appropriate cpp functions like `Gio::...`

eneshecan commented on src/Settings.cpp:
```diff
@@ -15,6 +16,26 @@ namespace
+
+    std::string getDesktopFilePath()
+    {
+        std::string possiblePath = "/usr/local/share/applications/whatsapp-for-linux.desktop";
```
Please use `auto`

eneshecan commented on src/Settings.cpp:
```diff
@@ -15,6 +16,26 @@ namespace
+        if(g_file_query_exists(gfilePath, nullptr))
+            return possiblePath;
+
+        return "";
```
Please use default empty initializer `{}`

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +176,23 @@ bool Settings::getStartInTray() const
+    auto dfPath = getDesktopFilePath();
+    if(dfPath.empty())
+    {
+        std::cerr << "setAutoStart() failed. Couldn't find .desktop file" << std::endl;
```
Please start with the class name in the print message i.e.  `Settings: Failed to find desktop file`.

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +176,23 @@ bool Settings::getStartInTray() const
+    }
+
+    auto homeDir = std::getenv("HOME");
+    auto gfileDestPath = strcat(homeDir, "/.config/autostart/whatsapp-for-linux.desktop");
```
Use the `CONFIG_DIR` variable in the free namespace above to get the `config` directory.

berkeyvx replied:
`CONFIG_DIR` returns `/home/{username}/.config/whatsapp-for-linux`. What i need is `/home/{username}/.config/autostart/whatsapp-for-linux.desktop`. I need to remove `/whatsapp-for-linux` from CONFIG_DIR, which is almost same as what i did.

Since `g_get_user_config_dir()` returns config directory of user and we use that directory in different parts of code we can change `CONFIG_DIR` to `std::string{g_get_user_config_dir()}` only and can make a new variable for `std::string{g_get_user_config_dir()} + "/whatsapp-for-linux";` What do you think about it?

eneshecan replied:
Then, we can make `CONFIG_DIR` to return the `.config` dir and have a separate variable for `CONFIG_APP_DIR`. Then you can use that variable instead of calling `getenv`.

eneshecan reviewed (commented):
Looks promising. Beware that we also need a switch in the header bar for this, please also implement that and related functions like `getAutoStart` etc.

berkeyvx commented:
> Looks promising. Beware that we also need a switch in the header bar for this, please also implement that and related functions like `getAutoStart` etc.

I thought we are going to use `Start In Tray` option to start aplication at tray during booting. Now i realized that we need another button. I will do asap. I think i fixed all of it. Can you review again?

Sorry for delay btw.

berkeyvx commented:
I have added `Settings::get/setAutostart` to save status of button to Settings file. Hence i renamed `setAutostart` function to `autostart`. 

When `autostart_switch` toggled, it save its status to `Settings` file. Now In `MainWindow` dtor it calls `Settings::getInstance().autostart()`. We can move `autostart` function to somewhere else, maybe private method of MainWindow but this requires to move `std::string getDesktopFilePath()` function too which is in anonymous namespace in Settings.cpp file.

If i did not miss anything, feature is now complete and work as expected.

eneshecan commented:
Thanks for the effort! I'll take a look soon.

eneshecan commented on resource/ui/MainWindow.ui:
```diff
@@ -103,7 +103,7 @@
                 <property name="visible">True</property>
                 <property name="can-focus">False</property>
-                <property name="label" translatable="yes">Close To Tray</property>
+                <property name="label" translatable="yes">          Close To Tray          </property>
```
Why? Please don't use spaces at the beginning and end, if needed use the appropriate policies.

eneshecan commented on resource/ui/MainWindow.ui:
```diff
@@ -141,7 +141,7 @@
                 <property name="visible">True</property>
                 <property name="can-focus">False</property>
-                <property name="label" translatable="yes">Start In Tray</property>
+                <property name="label" translatable="yes">           Start In Tray           </property>
```
Why? Please don't use spaces at the beginning and end, if needed use the appropriate policies.

eneshecan commented on resource/ui/MainWindow.ui:
```diff
@@ -169,6 +169,44 @@
+              <object class="GtkLabel" id="autostart_label">
+                <property name="visible">True</property>
+                <property name="can-focus">False</property>
+                <property name="label" translatable="yes">Autostart with System</property>
```
`Autostart` would be sufficient

eneshecan commented on src/MainWindow.cpp:
```diff
@@ -72,10 +76,16 @@ MainWindow::MainWindow(BaseObjectType* cobject, Glib::RefPtr<Gtk::Builder> const
 
+MainWindow::~MainWindow()
+{
+    Settings::getInstance().autostart();
```
Please rename the function as `setAutoStart`

berkeyvx replied:
I removed this completely. With moving autostart to setAutostart,  We do not need dtor. This was needed when performing autostart only once.

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +177,42 @@ bool Settings::getStartInTray() const
+    return autostart;
+}
+
+void Settings::autostart()
```
Move this logic into `setAutoStart`

eneshecan commented on src/Settings.hpp:
```diff
@@ -22,6 +22,9 @@ class Settings
         bool   getStartInTray() const;
+        void   setAutostart(bool autostart);
+        bool   getAutostart();
+        void   autostart();
```
Move this logic into `setAutoStart` so that the file is copied/removed everytime `setAutoStart` is called.

berkeyvx replied:
This operation could be performed only once while program is closing. That is why we call this function at `MainWindow` dtor. Executing each time when user toggle the button is unnecessary. But to keep the structure of Setting class i can change it. What do you think about it?

eneshecan replied:
Well, setAutoStart function itself should do what its name refers, it should check if the file exists or not and make the appropriate copy/remove. If you think it will be called too many times when the switch is toggled which is possible, we can only call this function in the dtor of MainWindow. Otherwise, I don't think that will be a big problem since it's only one file and it's under user's responsibility to not abuse it.

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +177,42 @@ bool Settings::getStartInTray() const
+    auto gfileSrc = Gio::File::create_for_path(dfPath);
+    auto gfileDest = Gio::File::create_for_path(gfileDestPath);
+
+    if(Settings::getInstance().getAutostart())
```
Do the appropriate checks to see if the file already exists before copying to prevent unnecessary operation.

berkeyvx replied:
I already do that here

```cpp
    if(dfPath.empty())
    {
        std::cerr << "Settings: Failed to find desktop file" << std::endl;
        return;
    }
 ```

eneshecan commented on src/Settings.cpp:
```diff
@@ -155,3 +177,42 @@ bool Settings::getStartInTray() const
+    if(Settings::getInstance().getAutostart())
+        gfileSrc->copy(gfileDest);
+    else
+        gfileDest->remove();
```
Do the appropriate checks to see if the file is already nonexistent before removing it to prevent unnecessary operation.

eneshecan replied:
I also realized that we don't need to keep this setting in the keyfile since the existense of the desktop file already corresponds to the value of it.

berkeyvx replied:
How to set state of Autostart switch?
```cpp
autostartSwitch->set_state(Settings::getInstance().getAutostart()); // in MainWindow
```
```cpp
    std::string getDesktopFilePath()
    {
        auto possiblePath = std::string("/usr/local/share/applications/whatsapp-for-linux.desktop");
        auto gfilePath = Gio::File::create_for_path(possiblePath);
        if(gfilePath->query_exists())
            return possiblePath;

        possiblePath = "/usr/share/applications/whatsapp-for-linux.desktop";
        gfilePath = Gio::File::create_for_path(possiblePath);
        if(gfilePath->query_exists())
            return possiblePath;

        possiblePath = "/snap/whatsapp-for-linux/current/share/applications/whatsapp-for-linux.desktop";
        gfilePath = Gio::File::create_for_path(possiblePath);

        if(gfilePath->query_exists())
            return possiblePath;

        return {};
    }
 ```

I can move this function body to `getAutostart` and write MainWindow autostartSwitch like this
```cpp
autostartSwitch->set_state(!Settings::getInstance().getAutostart().empty()); 
```

berkeyvx replied:
I can do something like this.

```cpp
bool Settings::getAutostart()
{
    auto gfileDestPath = CONFIG_DIR + "/autostart/whatsapp-for-linux.desktop";
    auto gfileDesktop = Gio::File::create_for_path(gfileDestPath);
    return gfileDesktop->query_exists();
}
```

Maybe we should move `CONFIG_DIR + "/autostart/whatsapp-for-linux.desktop";` to namespace because its constant and i used it in setAutostart too.

eneshecan reviewed (commented):
Looks much better :+1:

berkeyvx commented:
There is so much way to implement this so before making any other commit, lets decide how to do it. 
```cpp
    auto const CONFIG_DIR                            = std::string{g_get_user_config_dir()};
    auto const CONFIG_APP_DIR                        = CONFIG_DIR +  "/whatsapp-for-linux";
    auto const CONFIG_FILE_PATH                      = CONFIG_APP_DIR + "/settings.conf";
    auto const AUTOSTART_DESKTOP_FILE_PATH           = CONFIG_DIR + "/autostart/whatsapp-for-linux.desktop";
    constexpr auto const GROUP_GENERAL               = "General";
    constexpr auto const GROUP_NETWORK               = "Network";
    constexpr auto const POSSIBLE_DESKTOP_FILE_PATHS = {
            "/usr/local/share/applications/whatsapp-for-linux.desktop",
            "/usr/share/applications/whatsapp-for-linux.desktop",
            "/snap/whatsapp-for-linux/current/share/applications/whatsapp-for-linux.desktop"
    };
```
    
 I think It is better to move `possible desktop file paths` and `autostart desktop file path` to namespace. Easier to modify in the future and maintain.
 
 So i rewrite `getDesktopFilePath`

```cpp
    std::string getDesktopFilePath()
    {
        auto it = std::find_if(POSSIBLE_DESKTOP_FILE_PATHS.begin(), POSSIBLE_DESKTOP_FILE_PATHS.end(),[](auto const& elem){
            auto gfile = Gio::File::create_for_path(elem);
            return gfile->query_exists();
        });

        if (it == POSSIBLE_DESKTOP_FILE_PATHS.end())
            return {};
        return *it;
    }
```

Gave up on saving `Autostart` to keyfile, rewrote `getAutostart` (need this to set state of Autostart switch)
```cpp
bool Settings::getAutostart()
{
    auto gfileDesktop = Gio::File::create_for_path(AUTOSTART_DESKTOP_FILE_PATH);
    return gfileDesktop->query_exists();
}
```
`setAutostart` is same as before. What do you think about it?

eneshecan commented:
Great work! Thanks a lot.
